azkaban-common/src/main/java/azkaban/executor/ExecutionControllerUtils.java

```diff
@@ -231,9 +229,30 @@ static ExecutableFlow getFlowToRestart(final ExecutableFlow flow,
       restartedStatuses.add(Status.EXECUTION_STOPPED.name());
     }
 
+    // flow can only retry if custom-retry-times not reach the limit
+    // and for the case where the retry-times not defined, default to 1
+    int flowMaxRetryLimit = 0;
+    if (flowParams.containsKey(FlowParameters.FLOW_PARAM_MAX_RETRIES)){
+      flowMaxRetryLimit = Integer.parseInt(flowParams.get(FlowParameters.FLOW_PARAM_MAX_RETRIES));
+    }
+    else if (!restartedStatuses.isEmpty()){
+      logger.info(String.format("ExecutableFlow: %s has `%s=%s but `%s` not set, "
+              + "default to retry once",
+          FlowParameters.FLOW_PARAM_ALLOW_RESTART_ON_STATUS,
+          FlowParameters.FLOW_PARAM_MAX_RETRIES,
+          flow.getExecutionId(), restartedStatuses));
+      flowMaxRetryLimit = 1;
```

Comment:
Could you please check the order of the parameters passed to the format method? Shouldn't execution id be the first one?